### PR TITLE
feat(l1): fix snap sync + add healing

### PR DESCRIPTION
**Pull request by fmoletta**
**Motivation**
Fix snap sync logic:
Instead of rebuilding all block's state via snap, we select a pivot block (sync head - 64) fetch its state via snap and then execute all blocks after it
Add Healing phase

Missing from this PR:
- Reorg handling
- Fetching receipts

**Description**

Closes #1455

### CODE REVIEW

File: crates/networking/docs/Sync.md [hunk 1]
```diff
@@ -2,23 +2,33 @@
...
-The next two steps are performed in parallel:
-On one side, blocks and receipts for all fetched headers are fetched via p2p and stored.
+A snap sync cycle begins by fetching all the block headers (via eth p2p) between the current head (latest canonical block) and the sync head (block hash sent by a forkChoiceUpdate).
 
-On the other side, the state is reconstructed via p2p snap requests. Our current implementation of this works as follows:
-We will spawn two processes, the `bytecode_fetcher` which will remain active and process bytecode requests in batches by requesting bytecode batches from peers and storing them, and the `fetch_snap_state` process, which will iterate over the fetched headers and rebuild the block's state via `rebuild_state_trie`.
+We will then fetch the block bodies from each header and at the same time select a pivot block (sync head - 64) and start rebuilding its state via snap p2p requests, if the pivot were to become stale during this rebuild we will select a newer pivot (sync head) and restart it.
```
Comment (mpaulucci) [inline crates/networking/docs/Sync.md:7]: if the pivot becomes stale, do you discard all the state that you have downloaded so far?

File: crates/networking/docs/Sync.md [hunk 2]
```diff
@@ -2,23 +2,33 @@
...
 
-`rebuild_state_trie` will spawn a `storage_fetcher` process (which works similarly to the `bytecode_fetcher` and is kept alive for the duration of the rebuild process), it will open a new state trie and will fetch the block's accounts in batches and for each account it will: send the account's code hash to the `bytecode_fetcher` (if not empty), send the account's address and storage root to the `storage_fetcher` (if not empty), and add the account to the state trie. Once all accounts are processed, the final state root will be checked and committed.
+After we fully rebuilt the pivot state and fetched all the block bodies we will fetch and store the receipts for the range between the current head and the pivot (including it), and at the same time store all blocks in the same range and execute all blocks after the pivot (like in full sync).
```
Comment (mpaulucci) [inline crates/networking/docs/Sync.md:9]: I've seen in other clients that block bodies and receipts are downloaded in parallel (maybe after downloading the the state?). Any reason we do this like this here or is it just for simplicity?

File: crates/networking/docs/Sync.md [hunk 3]
```diff
@@ -2,23 +2,33 @@
...
 
-(Not implemented yet) When `fetch_snap_state` runs out of available state (aka, the state we need to fetch is older than 128 blocks and peers don't provide it), it will begin the `state_healing` process.
 This diagram illustrates the process described above:
 
-![snap_sync](/crates/networking/docs/diagrams/snap_sync.jpg)
+![snap_sync](/crates/networking/docs/diagrams/snap_sync.jpg).
```
Comment (mpaulucci) [inline crates/networking/docs/Sync.md:13]: nit: SendCodeHahses (Leafs) has a typo

Comment (mpaulucci) [reply crates/networking/docs/Sync.md:13]: Also Sapwn

File: crates/networking/docs/Sync.md (hunk 1 above)
Comment (fmoletta) [reply crates/networking/docs/Sync.md:7]: No, we still keep the state so we don't need to download it again

File: crates/networking/docs/Sync.md (hunk 3 above)
Comment (fmoletta) [reply crates/networking/docs/Sync.md:13]: Nice catch!

File: crates/networking/p2p/peer_channels.rs [hunk 4]
```diff
@@ -318,4 +320,112 @@ impl PeerChannels {
         }
         Some((storage_keys, storage_values, should_continue))
     }
+
+    /// Requests state trie nodes given the root of the trie where they are contained and their path (be them full or partial)
+    /// Returns the nodes or None if:
+    /// - There are no available peers (the node just started up or was rejected by all other nodes)
+    /// - The response timed out
+    /// - The response was empty or not valid
```
Comment (mpaulucci) [inline crates/networking/p2p/peer_channels.rs:328]: Haven't read the implementation for this, but it's not clear what happens if some nodes are found and others have issues (partial success). Does it return `None` or the nodes that are valid? Seems like the comment should make this clear

Comment (fmoletta) [reply crates/networking/p2p/peer_channels.rs:328]: If a node is invalid then the whole response would be invalid, there are no partial successes.

File: crates/networking/p2p/sync.rs [hunk 5]
```diff
@@ -114,48 +128,57 @@ impl SyncManager {
...
-                    .into_iter()
-                    .zip(all_block_hashes.into_iter())
-                {
-                    latest_block_number = header.number;
-                    store.set_canonical_block(header.number, hash)?;
-                    store.add_block_header(hash, header)?;
+                let mut pivot_idx = if all_block_headers.len() > MIN_FULL_BLOCKS {
+                    all_block_headers.len() - MIN_FULL_BLOCKS
+                } else {
+                    all_block_headers.len() - 1
```
Comment (mpaulucci) [inline crates/networking/p2p/sync.rs:141]: it's a bit weird to do snap sync if there is less than 64 total blocks. Can't we default to full sync here or would this increase complexity? If it's not trivial I would not do it.

File: crates/networking/p2p/sync.rs [hunk 6]
```diff
@@ -114,48 +128,57 @@ impl SyncManager {
...
+                };
+                let mut pivot_root = all_block_headers[pivot_idx].state_root;
+                let mut pivot_number = all_block_headers[pivot_idx].number;
+
+                let mut stale_pivot =
+                    !rebuild_state_trie(pivot_root, self.peers.clone(), store.clone()).await?;
+                // If the pivot became stale, set a further pivot and try again
+                if stale_pivot && pivot_idx != all_block_headers.len() - 1 {
+                    warn!("Stale pivot, switching to newer head");
+                    pivot_idx = all_block_headers.len() - 1;
```
Comment (mpaulucci) [inline crates/networking/p2p/sync.rs:151]: shouldn't you update `all_block_headers` before choosing a new pivot? I imagine the state_pivot could happen after some time and there are a lot more headers.

File: crates/networking/p2p/sync.rs [hunk 7]
```diff
@@ -114,48 +128,57 @@ impl SyncManager {
...
-                let mut latest_block_number = 0;
-                for (header, hash) in all_block_headers
-                    .into_iter()
-                    .zip(all_block_hashes.into_iter())
-                {
-                    latest_block_number = header.number;
-                    store.set_canonical_block(header.number, hash)?;
-                    store.add_block_header(hash, header)?;
+                let mut pivot_idx = if all_block_headers.len() > MIN_FULL_BLOCKS {
+                    all_block_headers.len() - MIN_FULL_BLOCKS
```
Comment (mpaulucci) [inline crates/networking/p2p/sync.rs:139]: Does it make sense to have all block headers in memory? For mainnet it would be ~21M. If it's a first implementation it's fine, we can optimize later

Comment (mpaulucci) [reply crates/networking/p2p/sync.rs:139]: But maybe I would just do something like `pivot_idx = head_block.number - MIN_FULL_BLOCKS`

File: crates/networking/p2p/sync.rs (hunk 6 above)
Comment (mpaulucci) [reply crates/networking/p2p/sync.rs:151]: why are you choosing `pivot_idx = all_block_headers.len() - 1;`? It will only move forward by 63 blocks.

File: crates/networking/p2p/sync.rs [hunk 8]
```diff
@@ -114,48 +128,57 @@ impl SyncManager {
...
+                    pivot_number = all_block_headers[pivot_idx].number;
+                    stale_pivot =
+                        !rebuild_state_trie(pivot_root, self.peers.clone(), store.clone()).await?;
                 }
-                // If all processes failed then they are likely to have a common cause (such as unaccessible storage), so return the first error
-                for result in set.join_all().await {
-                    result?;
+                if stale_pivot {
+                    warn!("Stale pivot, aborting sync");
+                    return Ok(());
```
Comment (mpaulucci) [inline crates/networking/p2p/sync.rs:159]: return Error here?

File: crates/networking/p2p/sync.rs [hunk 9]
```diff
@@ -114,48 +128,57 @@ impl SyncManager {
...
+                // For all blocks before the pivot: Store the bodies and fetch the receipts
+                // For all blocks after the pivot: Process them fully
+                // let store_receipts_handle = tokio::spawn(store_receipts(all_block_hashes[pivot_index..]));
+                for (hash, (header, body)) in all_block_hashes.into_iter().zip(
+                    all_block_headers
+                        .into_iter()
+                        .zip(all_block_bodies.into_iter()),
+                ) {
+                    if header.number <= pivot_number {
+                        store.set_canonical_block(header.number, hash)?;
```
Comment (mpaulucci) [inline crates/networking/p2p/sync.rs:172]: I think the only one calling `set_canonical_block` should be the apply forkchoice function.

File: crates/networking/p2p/sync.rs [hunk 10]
```diff
@@ -114,48 +128,57 @@ impl SyncManager {
...
+                        store.add_block(Block::new(header, body))?;
+                    } else {
+                        store.set_canonical_block(header.number, hash)?;
+                        store.update_latest_block_number(header.number)?;
```
Comment (mpaulucci) [inline crates/networking/p2p/sync.rs:176]: idem, this is a job of the apply forkchoice update function, here we just store the block

File: crates/networking/p2p/sync.rs (hunk 6 above)
Comment (fmoletta) [reply crates/networking/p2p/sync.rs:151]: We are not fetching any more headers as we don't know about any later block aside from the sync root, the only way to sync further would be to wait for a next fork choice update.

File: crates/networking/p2p/sync.rs (hunk 7 above)
Comment (fmoletta) [reply crates/networking/p2p/sync.rs:139]: It made sense for small tests as we could store the full block once we had the body too, but yes, we wouldn't be able to do it with the full ethereum state

File: crates/networking/p2p/sync.rs (hunk 8 above)
Comment (fmoletta) [reply crates/networking/p2p/sync.rs:159]: I looked at the error handling on geth and they don't return an error from the main sync process.
The sync process happenes on its own, there is no one we need to reply to with the status of the sync, only the owner/user of the node itself (via tracing) so it doesn't make much sense to return an error unless we want to shut down the node entirely

File: crates/networking/p2p/sync.rs (hunk 9 above)
Comment (fmoletta) [reply crates/networking/p2p/sync.rs:176]: The problem is that we already left the fork choice function by the time the sync ends. 
Syncs are only triggered by forkChoiceUpdates so I think it makes sense to consider the sync root as the new head of the canonical chain.
Also the ethereum/sync hive test expects the latest block to be set after triggering the sync with a single forkChoiceUpdate

File: crates/networking/p2p/sync.rs (hunk 5 above)
Comment (fmoletta) [reply crates/networking/p2p/sync.rs:141]: Id say this is a very rare edge case, we only perform snap sync if:
A) We are just starting up the node, in which case it is very rare that the chain we sync to would have less than 64 blocks in a non-test case
B) A re-org threw us below the latest pivot, in which case it would also be very rare that the new canonical chain has only 64 blocks or less

**Review by mpaulucci — APPROVED** [review]

File: crates/networking/p2p/sync.rs (hunk 8 above)
Comment (fmoletta) [reply crates/networking/p2p/sync.rs:159]: Currently, the rpc context has a mutex reference to the SyncManager, if it can't get hold of the Mutex then it assumes there is an active sync process and returns the syncing status.
